scene/gui/progress_bar.cpp

```diff
@@ -62,15 +62,55 @@ void ProgressBar::_notification(int p_what) {
 			Color font_color = get_theme_color(SNAME("font_color"));
 
 			draw_style_box(bg, Rect2(Point2(), get_size()));
+
 			float r = get_as_ratio();
-			int mp = fg->get_minimum_size().width;
-			int p = r * (get_size().width - mp);
-			if (p > 0) {
-				if (is_layout_rtl()) {
-					draw_style_box(fg, Rect2(Point2(p, 0), Size2(fg->get_minimum_size().width, get_size().height)));
-				} else {
-					draw_style_box(fg, Rect2(Point2(0, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
-				}
+
+			switch (mode) {
+				case FILL_BEGIN_TO_END: {
+					int mp = fg->get_minimum_size().width;
+					int p = round(r * (get_size().width - mp));
+
+					if (p > 0) {
+						if (is_layout_rtl()) {
+							int p_remaining = round((1.0 - r) * (get_size().width - mp));
+							draw_style_box(fg, Rect2(Point2(p_remaining, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
+						} else {
+							draw_style_box(fg, Rect2(Point2(0, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
+						}
+					}
+				} break;
+				case FILL_END_TO_BEGIN: {
+					int mp = fg->get_minimum_size().width;
+					int p = round(r * (get_size().width - mp));
+
+					if (p > 0) {
+						if (is_layout_rtl()) {
+							draw_style_box(fg, Rect2(Point2(0, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
+						} else {
+							int p_remaining = round((1.0 - r) * (get_size().width - mp));
+							draw_style_box(fg, Rect2(Point2(p_remaining, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
+						}
+					}
+				} break;
```

Comment:
Could maybe be simplified like this (untested):
```suggestion
				case FILL_BEGIN_TO_END:
				case FILL_END_TO_BEGIN: {
					int mp = fg->get_minimum_size().width;
					int p = round(r * (get_size().width - mp));

					// We want FILL_BEGIN_TO_END to map to right to left when UI layout is RTL,
					// and left to right otherwise. And likewise for FILL_END_TO_BEGIN.
					// These two checks achieve that.
					bool right_to_left = is_layout_rtl() ? (mode == FILL_BEGIN_TO_END) : (mode == FILL_END_TO_BEGIN);

					if (p > 0) {
						if (right_to_left) {
							int p_remaining = round((1.0 - r) * (get_size().width - mp));
							draw_style_box(fg, Rect2(Point2(p_remaining, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
						} else {
							draw_style_box(fg, Rect2(Point2(0, 0), Size2(p + fg->get_minimum_size().width, get_size().height)));
						}
					}
				} break;
```